### PR TITLE
Post mortem for the incident involving filenames with slashes

### PR DESCRIPTION
**Pull request by danidoni**
Post mortem for the Incident https://etherpad.opensuse.org/p/INCIDENT-verify_the_file_exists%2315834

### CODE REVIEW

File: _posts/deployments/2023-03-21-post-mortem.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+There was a service degradation of our reference server.
+
+<!--
+  What happened, for how long and who was impacted by it?
+  For customers to be able to identify if their problem is related to this post-mortem or not.
+-->
+
+
+After a deployment of build.opensuse.org on Tuesday, March 19, 10:36 UTC,
+accessing the changes on some Submit Requests showed errors to the user. Specifically, those requests that had changes on files containing a slash symbol in its path (e.g: buildah-1.35.1.tar.xz/CHANGELOG.md).
```
Comment (hennevogel) [inline _posts/deployments/2023-03-21-post-mortem.md:17]: Which error? :)

File: _posts/deployments/2023-03-21-post-mortem.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+We want to give you some insight into what happened and what we are doing to avoid similar problems in the future.
+
+## Detection
+<!--
+  How did we get alerted that the problem happened?
+  To demonstrate to customers how we are (hopefully automatically) alerted about problems.
+-->
+
+We were contacted by a user in our #help-obs Slack channel, from there, we saw that there were errors in our error collector since 10:36 UTC.
```
Comment (hennevogel) [inline _posts/deployments/2023-03-21-post-mortem.md:29]: Nobody outside of SUSE knows this exists.

File: _posts/deployments/2023-03-21-post-mortem.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+## Root Cause
+<!--
+  What was it and what triggered it?
+  For customers and community to understand what happened technically.
+-->
+
+The root cause was a refactoring we introduced on the file viewing capabilities of OBS. It didn't accept slashes in file names.
```
Comment (hennevogel) [inline _posts/deployments/2023-03-21-post-mortem.md:37]: It also did not have a proper spec...

File: _posts/deployments/2023-03-21-post-mortem.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+N/A
+
+## Lessons Learned
+<!--
+  Describe what went well, what went wrong and where we go lucky during the resolution of this problem.
+-->
+
+- We were not aware the code before the refactoring was producing filenames with slashes, and happily accepting them.
+- While refactoring we have to be very careful into ensuring OBS behaves the same before and after the changes.
```
Comment (hennevogel) [inline _posts/deployments/2023-03-21-post-mortem.md:62]: No we don't need to be careful, we need comprehensive specs before doing a refactoring...

Comment (danidoni) [reply _posts/deployments/2023-03-21-post-mortem.md:62]: Right

File: _posts/deployments/2023-03-21-post-mortem.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+The root cause was a refactoring we introduced on the file viewing capabilities of OBS. That refactoring introduced the rejection of file names containing slashes. Not having a comprehensive test coverage prior to that refactoring made the detection of that flaw impossible.
+
+## Resolution
+<!--
+  How did you resolve or work around this problem?
+  For customers and community to understand what happened technically.
+-->
+
+We corrected the code to avoid linking to filenames with slashes.
```
Comment (hennevogel) [inline _posts/deployments/2023-03-21-post-mortem.md:45]: We also introduced an automated test for this case.

Comment (danidoni) [reply _posts/deployments/2023-03-21-post-mortem.md]: Right

File: _posts/deployments/2023-03-21-post-mortem.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+There was a service degradation of our reference server.
+
+<!--
+  What happened, for how long and who was impacted by it?
+  For customers to be able to identify if their problem is related to this post-mortem or not.
+-->
+
+
+After a deployment of build.opensuse.org on Tuesday, March 19, 10:36 UTC,
+accessing the changes on some Submit Requests showed [this error](https://github.com/openSUSE/open-build-service/issues/15834) to the user. Specifically, those requests that had changes on files containing a slash symbol in its path (e.g: buildah-1.35.1.tar.xz/CHANGELOG.md).
```
[the remark this answers is not in the document]
Comment (danidoni) [reply _posts/deployments/2023-03-21-post-mortem.md]: I've added the original error

File: _posts/deployments/2024-03-21-post-mortem.md [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+<!--
+  What happened, for how long and who was impacted by it?
+  For customers to be able to identify if their problem is related to this post-mortem or not.
+-->
+
+
+After a deployment of build.opensuse.org on Tuesday, March 19, 10:36 UTC,
+accessing the changes on some Submit Requests showed this error to the user:
+
+![](/images/posts/2024-03-21-post-mortem/error.png)
```
Comment (hennevogel) [inline _posts/deployments/2024-03-21-post-mortem.md:19]: This is another issue. I guess the people just saw the 500 page no?

File: _posts/deployments/2024-03-21-post-mortem.md [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+  Describe what went well, what went wrong and where we go lucky during the resolution of this problem.
+-->
+
+- We were not aware the code before the refactoring was producing filenames with slashes, and happily accepting them.
+- Before refactoring anything, we need a comprehensive test coverage to ensure we are not introducing errors.
+
+## Timeline (All Time in UTC)
+
+* 2024-03-19 10:36 First appearance of the error in our error collector.
+* 2024-03-19 12:37 Some user detected an issue on our #help-obs Slack channel.
```
Comment (hennevogel) [inline _posts/deployments/2024-03-21-post-mortem.md:71]: Slack is unknown to the public

File: _posts/deployments/2024-03-21-post-mortem.md (hunk 7 above)
Comment (danidoni) [reply _posts/deployments/2024-03-21-post-mortem.md:19]: Hmmm, now that I look into it, this is only shown for users parcipating in the beta. Let me find what user see when not participating in the beta.

File: _posts/deployments/2024-03-21-post-mortem.md (hunk 8 above)
Comment (danidoni) [reply _posts/deployments/2024-03-21-post-mortem.md:75]: True again

File: _posts/deployments/2024-03-21-post-mortem.md (hunk 7 above)
Comment (danidoni) [reply _posts/deployments/2024-03-21-post-mortem.md]: I've added the errors messages show to both beta and non beta users

Comment (hennevogel) [reply _posts/deployments/2024-03-21-post-mortem.md]: I see. We don't need to mention the beta program. It's supposed to break :)

Comment (danidoni) [reply _posts/deployments/2024-03-21-post-mortem.md]: What do you mean "it's supposed to break"?

Comment (danidoni) [reply _posts/deployments/2024-03-21-post-mortem.md]: You mean it's not worth showing the error that appears to users under the beta program?

Comment (hennevogel) [reply _posts/deployments/2024-03-21-post-mortem.md]: Yes it's not worth mentioning what happened in the beta program at all...

Comment (danidoni) [reply _posts/deployments/2024-03-21-post-mortem.md:19]: Fixed

**Review by hennevogel — APPROVED** [review]

**Pull request merged**